### ISSUE TITLE
JOSS REVIEW - Code

### ISSUE THREAD

**Issue by rsenne** [johnsonandjohnson/open_nipals#12]
# Code Review for `open_nipals`

Here is a non-exhaustive list of things I've written notes on in the past few weeks. As a note, I haven't developed a lot of python code in recent history, and much of this I didn't personally test, but these are things I noticed that I thought _could_ be problems, so if there is overreach here I apologize. Let me know if you have questions--I hope this is helpful.

### 1) Reliance on private scikit-learn classes
**Problem.** `NipalsPCA` subclasses `sklearn.decomposition._base._BasePCA` and `NipalsPLS` subclasses `sklearn.cross_decomposition._pls._PLS`. These underscored APIs are private and unstable across releases. The classes also do not call `super().__init__`, and several standard scikit-learn attributes are not exposed.

**Recommendation.**
- Inherit from `BaseEstimator` + `TransformerMixin` (and `RegressorMixin` for PLS) instead of private classes.
- Expose expected attributes and aliases:
  - PCA: `components_`, `explained_variance_`, `explained_variance_ratio_`, `singular_values_`, `mean_`, `n_features_in_`, `n_iter_`.
  - PLS: `x_weights_`, `x_loadings_`, `y_weights_`, `y_loadings_`, `x_scores_`, `y_scores_`, `coef_`, `n_iter_`.
- Use `validate_data` and `check_is_fitted` for input validation and fit checks.

---

### 2) Mean-centering semantics
**Problem.** `mean_centered` triggers warnings but the code does not store or consistently apply centering. `fit` does not compute `mean_`; `transform`/`inverse_transform` do not reliably subtract/add the mean.

**Recommendation.**
- Choose a clear contract and enforce it:
  - **Option A:** Always center in `fit` (store `mean_`); center in `transform`; add back in `inverse_transform`.
  - **Option B:** Require pre-centered data and assert it; document clearly.

---

### 3) Bug in `NipalsPLS.calc_imd`
**Problem.** When `input_array` is provided, you compute `scores = self.transform(...)` but later use `input_scores` to form \(T^2\), which may be `None`.

**Fix.**
```python
if input_array is not None:
    S = self.transform(input_x=input_array)
else:
    S = input_scores
# ...
out_imd = np.sum((S - fit_means)**2 / fit_vars, axis=1).reshape(-1, 1)
````

---

### 4) Numerical fragility: direct matrix inverses

**Problem.** Direct inverses are used in projection/conditional-mean paths (e.g., $(L^\top L)^{-1}$, $S_{22}^{-1}$). These can be singular/ill-conditioned with missing data or collinearity.

**Recommendation.**

* Replace `np.linalg.inv` with `np.linalg.pinv` or regularized solves (e.g., Tikhonov). Detect rank deficiency and warn.
* Example:

```python
A = loadings[not_null, :num_lvs]
denom = np.linalg.pinv(A.T @ A)  # or (A.T @ A + lam * I)^-1
```

---

### 5) `conditionalMean` refits a full model inside `transform`

**Problem.** For `fit_cols > n_components`, a new full `NipalsPCA(fit_cols)` is fit during `transform`. This is expensive and silently changes the basis used for imputation.

**Recommendation.**

* Precompute and cache full-basis quantities during `fit` (behind a flag), or require a full model to be supplied by the caller.
* At minimum, memoize the full model to avoid repeated refits.

---

### 6) Hotelling’s $T^2$: diagonal-only variance

**Problem.** $T^2$ is computed with per-component variances (diagonal assumption). With missing data and NIPALS updates, score covariance can be non-diagonal.

**Recommendation.**

* Provide `covariance={'diag','full'}`. For `full`, compute $T^2=(t-\mu)^\top\Sigma^{-1}(t-\mu)$ with shrinkage (e.g., Ledoit–Wolf). Default can remain `diag` for speed but document the assumption.

---

### 7) DModX/QRes consistency between PCA and PLS

**Problem.** PCA computes DModX with row-wise observed-feature counts $K$; PLS uses a global $K$. SIMCA factors are mentioned but not justified.

**Recommendation.**

* Align formulas and cite them explicitly. If following SIMCA, expose `mode={'simca','raw'}` and document scaling and assumptions.
* Add tests comparing PCA vs. PLS diagnostics on the same synthetic data.

---

## Correctness and robustness

* **Convergence divide-by-zero.** Guard $\|t_{new}\|$ with an epsilon in the convergence denominator.
* **`_nan_mult` assumes column vector `y`.** For multi-column `y`, algebra/broadcasting is ambiguous. Either assert `y.shape[1] == 1` or implement the general case.
* **Normalization duplication.** If loadings are unit norm at fit time, dividing by `(loadings.T @ loadings)` again in `transform` is redundant. Ensure consistent normalization across NaN/no-NaN paths.
* **Warnings vs. prints.** Replace `print("max_iter...")` with `warnings.warn(...)` to respect user-configured warning filters.
* **Use `np.concatenate` instead of `np.append`.** `np.append` can flatten arrays and is slower for 2D data.

---

## API and UX

* Expose the standard scikit-learn attributes listed above; document any semantic differences (e.g., sign/normalization).
* `set_components` mutates model state and may be expensive. Document its cost and consider a `refit(n_components=...)` pathway that restarts cleanly.
* Determinism: support `random_state` and a `start={'col','svd','random'}` initializer.

---

## Performance

* Defaults (`max_iter=10_000`, `tol=1e-8/1e-10`) are strict. Provide rationale and guidance; consider looser defaults (e.g., `1e-6`) and early-stopping with patience.
* Reduce allocation churn in iterative loops by reusing buffers for residuals and differences on large problems.

---

## Testing (suggested additions)

1. Missingness patterns: MCAR vs. block-missing; extreme sparsity; rows with all-NaN in `Y` with `force_include=True/False`.
2. Degenerate cases: rank-deficient `X`; duplicate columns; single feature; single sample.
3. Convergence: hitting `max_iter` emits a warning; zero-norm guards work.
4. Projection methods: compare `projection`/`conditionalMean` to a PPCA/EM reference on toy data; verify stability with `pinv`.
5. API parity: `inverse_transform(transform(X))` error bounds; `set_components` up/down consistent with a fresh refit.
6. Diagnostics: DModX/QRes consistent scaling across PCA/PLS on synthetic data.
---

## Small patch candidates

1. **Fix `NipalsPLS.calc_imd` variable handling**

```python
if input_array is not None:
    S = self.transform(input_x=input_array)
else:
    S = input_scores
# use S below
```

2. **Guard convergence division**

```python
den = max(np.linalg.norm(t_new), 1e-12)
conv_test = np.linalg.norm(t_old - t_new) / den
```

3. **Use pseudo-inverse in projection**

```python
A = loadings[not_null, :num_lvs]
scores[row, :] = (np.linalg.pinv(A.T @ A) @ (A.T @ input_array[row, not_null].T)).T
```

4. **Warnings instead of prints**

```python
if iter_count >= self.max_iter:
    warnings.warn(f"max_iter reached on LV {ind_lv}")
```

5. **Enforce column vector in `_nan_mult`**

```python
if y.ndim != 2 or y.shape[1] != 1:
    raise ValueError("_nan_mult expects y to have shape (m, 1)")
```

6. **Replace `np.append` with `np.concatenate`**

```python
scores = np.concatenate([self.fit_scores, np.zeros((n, n_add))], axis=1)
```

**Comment by nielsschlusser** [comment]
Major Issues:
1. We have changed the base classes from the private ones to the ones you suggested. We have also exposed some of the attributes you suggested. As our package is designed to operate after standardizing the data with `sklearn.StandardScaler`, we refrained from exposing `mean_ `.
2. For precisely that reason, we decided not to include an automatic mean centering step into our package, also, because in case anyone _really_ knows what they are doing, we would like to grant them the opportunity to use the package on non-mean-centered data. We have pointed out in the documentation that we _highly_ recommend to mean_center the data.
3. Thank you very much for pointing out the bug. We have fixed it.
4. According to your suggestion, we have replaced the direct matrix inverses by pseudo-inverses.
5. Thank you for pointing out the re-fit of the entire model in the `conditional_mean` case of `transform`. This was indeed a very costly and suboptimal implementation. We have replaced it by a double use of the `set_components()` method, see [code](https://github.com/johnsonandjohnson/open_nipals/blob/7d66eb590e8667b2c6f8462f7618dd249c046701/src/open_nipals/nipalsPCA.py#L379-L387). 
6. We have expanded the `calc_imd()` function by a `covariance` option, that can be either of `{'diag', 'full', 'ledoit_wolf'}`.
7. Thank you for pointing out the lack of consistency between `calc_oomd()` for PCA and PLS. Unfortunately, this inconsistency seems to originate from SIMCA itself. In order to maintain consistency with SIMCA, we left it as is.

**Comment by nielsschlusser** [comment]
Performance:
Default tolerance values of `1e-8/1e-10` are indeed too strict. We have set a less rigid default value of `1e-6`.

**Comment by rsenne** [comment]
LGTM. Thanks for making all of these changes!
